### PR TITLE
Add deepimagej config generation

### PR DESCRIPTION
**Pull request by constantinpape**
Write the deepimagej config. Still need to get the pixel size information somehow.
cc @esgomezm

### CODE REVIEW

**Comment by esgomezm** [comment]
Hi @constantinpape 

In the ZeroCostDL4Mic notebook for StarDist we directly ask the user for this value. By default the pixel size could be set to 1 pixel. 

In the rest of the notebooks, it's calculated using the following function:
```python
from PIL import Image 
def getPixelSizeTIFFmetadata(TIFFpath, display=False):
  with Image.open(TIFFpath) as img:
    meta_dict = {TAGS[key] : img.tag[key] for key in img.tag.keys()}


  # TIFF tags
  # https://www.loc.gov/preservation/digital/formats/content/tiff_tags.shtml
  # https://www.awaresystems.be/imaging/tiff/tifftags/resolutionunit.html
  ResolutionUnit = meta_dict['ResolutionUnit'][0] # unit of resolution
  width = meta_dict['ImageWidth'][0]
  height = meta_dict['ImageLength'][0]

  xResolution = meta_dict['XResolution'][0] # number of pixels / ResolutionUnit

  if len(xResolution) == 1:
    xResolution = xResolution[0]
  elif len(xResolution) == 2:
    xResolution = xResolution[0]/xResolution[1]
  else:
    print('Image resolution not defined.')
    xResolution = 1

  if ResolutionUnit == 2:
    # Units given are in inches
    pixel_size = 0.025*1e9/xResolution
  elif ResolutionUnit == 3:
    # Units given are in cm
    pixel_size = 0.01*1e9/xResolution
  else: 
    # ResolutionUnit is therefore 1
    print('Resolution unit not defined. Assuming: um')
    pixel_size = 1e3/xResolution

  if display:
    print('Pixel size obtained from metadata: '+str(pixel_size)+' nm')
    print('Image size: '+str(width)+'x'+str(height))
  
  return (pixel_size, width, height)
```

**Comment by constantinpape** [comment]
Thanks for sharing the script @esgomezm, but I fear that this is not very useful for us because we can't rely on having tif files as inputs; the (mandatory) input data are numpy arrays here.
I will go ahead and just add the pixel size as a parameter to the `build_model` function and propagate it to the deepimagej config; then you can pass it to the function.

**Comment by FynnBe** [comment]
incorporating pixel size into the spec (0.4.?) sounds like another hackathon item to me...

**Comment by constantinpape** [comment]
> incorporating pixel size into the spec (0.4.?) sounds like another hackathon item to me...

Yes, def. a point to discuss

**Comment by constantinpape** [comment]
Ok, I updated the deepimagej config function to include the pixel size and fix several issues.
There are still a couple open points:
- I don't understand the deepimagej binarize macro and how to set the threshold in it @esgomezm
- Does deepimagej support models with multiple inputs / outptus? If yes, how to specify the pre-post-processing for this in the config? Afaik there is just a single list of pre/postprocessing, which implies that this is applied to a single in/out tensors

Also, the tests are still failing due to some issues with sample inputs / outputs that I don't understand. But this may be fixed by #164, so we should merge that PR first and then merge main here.

**Comment by FynnBe** [comment]
> Also, the tests are still failing due to some issues with sample inputs / outputs that I don't understand. But this may be fixed by #164, so we should merge that PR first and then merge main here.

yes, this should be fixed by #164

File: bioimageio/core/build_spec/build_model.py [hunk 1]
```diff
@@ -258,6 +266,160 @@ def _get_dependencies(dependencies, root):
...
+
+    elif name == "zero_mean_unit_variance":
+        mode = kwargs["mode"]
+        if mode == "fixed":
+            macro = "fixed_zero_mean_unit_variance.ijm"
+            replace = {"paramMean": kwargs["mean"], "paramStd": kwargs["std"]}
+        else:
+            macro = "zero_mean_unit_variance.ijm"
+            replace = {}
+
```
Comment (FynnBe) [inline bioimageio/core/build_spec/build_model.py:281]:
```suggestion
    else:
        raise ValueError(name)

```
would save you the duplication of available macro names in the assert above

File: bioimageio/core/build_spec/build_model.py [hunk 2]
```diff
@@ -258,6 +266,160 @@ def _get_dependencies(dependencies, root):
...
+    # elif name == "binarize":
+    #     macro = "binarize.ijm"
+    #     replace = {"", kwargs["threshold"]}
+
+    macro = f"{name}.ijm"
+    url = f"https://raw.githubusercontent.com/deepimagej/imagej-macros/master/bioimage.io/{macro}"
+
+    path = os.path.join(export_folder, macro)
+    # TODO do we use requests?
+    with requests.get(url, stream=True) as r:
```
Comment (FynnBe) [inline bioimageio/core/build_spec/build_model.py:291]: let me update https://github.com/bioimage-io/core-bioimage-io-python/blob/74e0f4beafabdb5a43df0d082dd3731c7f968441/bioimageio/core/resource_io/utils.py#L267 to accept `local_path` to get a nice tqdm pbar and then use it here...

**Comment by constantinpape** [comment]
@FynnBe the tests for writing the deepimagej config is still failing with
```
 marshmallow.exceptions.ValidationError: {'sample_outputs': {0: ['Errors in all options for this field. Fix any of the following errors:', ['/tmp/bioimageio_cache/extracted_packages/8cfc68d44ac8c3a154290017d4151113277ca8a27e72a15b39fb4a0792b0ec79/sample_output_0.
tif'], ['expected relative path.']]}, 'sample_inputs': {0: ['Errors in all options for this field. Fix any of the following errors:', ['/tmp/bioimageio_cache/extracted_packages/8cfc68d44ac8c3a154290017d4151113277ca8a27e72a15b39fb4a0792b0ec79/sample_input_0.tif'], ['expected
 relative path.']]}} 
```
although sample_inputs/outputs are treated exactly the same as test_inputs/outputs in `build_model`.

File: bioimageio/core/build_spec/build_model.py [hunk 3]
```diff
@@ -477,6 +705,17 @@ def build_model(
...
+    # debugging: for some reason this causes a validation error for the sample_input / output,
+    # although they are passed in the same format as test inputs / outputs:
+    # (Pdb) model.sample_inputs
+    # [PosixPath('/tmp/bioimageio_cache/extracted_packages/efae734f36ff8634977c9174f01c854c4e2cfc799bc92751a4558a0edd963da6/sample_input_0.tif')]
+    # (Pdb) model.sample_outputs
+    # [PosixPath('/tmp/bioimageio_cache/extracted_packages/efae734f36ff8634977c9174f01c854c4e2cfc799bc92751a4558a0edd963da6/sample_output_0.tif')]
+    # (Pdb) model.test_inputs
+    # [PosixPath('/tmp/bioimageio_cache/extracted_packages/efae734f36ff8634977c9174f01c854c4e2cfc799bc92751a4558a0edd963da6/test_input.npy')]
+    # (Pdb) model.test_outputs
+    # [PosixPath('/tmp/bioimageio_cache/extracted_packages/efae734f36ff8634977c9174f01c854c4e2cfc799bc92751a4558a0edd963da6/test_output.npy')]
```
Comment (FynnBe) [inline bioimageio/core/build_spec/build_model.py:718]: should be fixed by #164 ; then this should be removed

**Review by FynnBe — COMMENTED** [review]
almost there 👍

File: bioimageio/core/build_spec/build_model.py (hunk 3 above)
Comment (FynnBe) [reply bioimageio/core/build_spec/build_model.py:718]: > the tests for writing the deepimagej config is still failing with

... or not.. looking into it

Comment (FynnBe) [reply bioimageio/core/build_spec/build_model.py:718]: the problem is that sample inputs/outputs are absolute and not relative paths...

Comment (constantinpape) [reply bioimageio/core/build_spec/build_model.py:718]: but test inputs / outptus are also absolute paths. Why does it work for them and not for sample inputs/outputs?

**Comment by FynnBe** [comment]
1942df0 contains bugs! not ready yet...

**Comment by constantinpape** [comment]
> - I don't understand the deepimagej binarize macro and how to set the threshold in it @esgomezm

It's `optimal_thresold`.

> - Does deepimagej support models with multiple inputs / outptus? If yes, how to specify the pre-post-processing for this in the config? Afaik there is just a single list of pre/postprocessing, which implies that this is applied to a single in/out tensors

It's in theory possible, but too complex for now; we will just throw an error for now and leave this for future iterations.

File: bioimageio/core/build_spec/build_model.py (hunk 1 above)
Comment (constantinpape) [reply bioimageio/core/build_spec/build_model.py:281]: But the error message is more meaningful in the other case. I will update a bit differently.

**Comment by FynnBe** [comment]
locally for me there remains an issue with sample_inputs/outputs serialization. Not sure what's going on there...

**Comment by FynnBe** [comment]
> although sample_inputs/outputs are treated exactly the same as test_inputs/outputs in build_model.

this turned out to be wrong... in the deepimagej they are treated differently... (this should be fixed now though

**Comment by FynnBe** [comment]
An update to bioimageio.spec should fix the remaining issue of serialization: https://github.com/bioimage-io/spec-bioimage-io/pull/294

**Comment by constantinpape** [comment]
Thanks @FynnBe now it all works.

**Pull request merged**